versioned/spi/src/main/java/org/projectnessie/versioned/EventsVersionStore.java

```diff
@@ -0,0 +1,223 @@
+ * limitations under the License.
+ */
+package org.projectnessie.versioned;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.Callable;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+import javax.annotation.Nonnull;
+import org.projectnessie.model.CommitMeta;
+import org.projectnessie.model.Content;
+import org.projectnessie.model.ContentKey;
+import org.projectnessie.model.MergeBehavior;
+import org.projectnessie.model.MergeKeyBehavior;
+import org.projectnessie.versioned.paging.PaginationIterator;
+
+/**
+ * A {@link VersionStore} wrapper that publishes results when a method is called that changes the
+ * catalog state.
+ */
+public class EventsVersionStore implements VersionStore {
+
+  private final VersionStore delegate;
+  private final Consumer<Result> resultSink;
+
+  /**
+   * Takes the {@link VersionStore} instance to enrich with events.
+   *
+   * @param delegate backing/delegate {@link VersionStore}.
+   * @param resultSink a consumer for results.
+   */
+  public EventsVersionStore(VersionStore delegate, Consumer<Result> resultSink) {
+    this.delegate = delegate;
+    this.resultSink = resultSink;
+  }
+
+  @Override
+  public CommitResult<Commit> commit(
+      @Nonnull @jakarta.annotation.Nonnull BranchName branch,
+      @Nonnull @jakarta.annotation.Nonnull Optional<Hash> referenceHash,
+      @Nonnull @jakarta.annotation.Nonnull CommitMeta metadata,
+      @Nonnull @jakarta.annotation.Nonnull List<Operation> operations,
+      @Nonnull @jakarta.annotation.Nonnull Callable<Void> validator,
+      @Nonnull @jakarta.annotation.Nonnull BiConsumer<ContentKey, String> addedContents)
+      throws ReferenceNotFoundException, ReferenceConflictException {
+    CommitResult<Commit> result =
+        delegate.commit(branch, referenceHash, metadata, operations, validator, addedContents);
+    resultSink.accept(result);
+    return result;
+  }
+
+  @Override
+  public MergeResult<Commit> transplant(
+      NamedRef sourceRef,
+      BranchName targetBranch,
+      Optional<Hash> referenceHash,
+      List<Hash> sequenceToTransplant,
+      MetadataRewriter<CommitMeta> updateCommitMetadata,
+      boolean keepIndividualCommits,
+      Map<ContentKey, MergeKeyBehavior> mergeKeyBehaviors,
+      MergeBehavior defaultMergeBehavior,
+      boolean dryRun,
+      boolean fetchAdditionalInfo)
+      throws ReferenceNotFoundException, ReferenceConflictException {
+    MergeResult<Commit> result =
+        delegate.transplant(
+            sourceRef,
+            targetBranch,
+            referenceHash,
+            sequenceToTransplant,
+            updateCommitMetadata,
+            keepIndividualCommits,
+            mergeKeyBehaviors,
+            defaultMergeBehavior,
+            dryRun,
+            fetchAdditionalInfo);
+    if (!dryRun) {
```

Comment:
Shouldn't this check `result.wasApplied()` instead to only emit the event, when the merge was actually persisted? (same for `merge()` below)